pallets/liquid-staking/src/lib.rs

```diff
@@ -684,12 +749,91 @@ pub mod pallet {
                 return Ok(());
             }
 
-            T::XCM::do_rebond(amount, Self::staking_currency()?, T::DerivativeIndex::get())?;
+            let query_id = T::XCM::do_rebond(
+                amount,
+                Self::staking_currency()?,
+                T::DerivativeIndex::get(),
+                Self::notify_placeholder(),
+            )?;
+            XcmRequests::<T>::insert(query_id, XcmRequest::Rebond { amount });
             Self::deposit_event(Event::<T>::Rebonding(amount));
 
             Ok(())
         }
 
+        #[require_transactional]
+        fn do_notification_received(
+            query_id: QueryId,
+            request: XcmRequest<T>,
+            res: Option<(u32, XcmError)>,
+        ) -> DispatchResult {
+            let executed = res.is_none();
+
+            match request {
+                XcmRequest::Bond { amount, .. } if executed => {
+                    MatchingPool::<T>::try_mutate(|p| -> DispatchResult {
+                        p.total_stake_amount = p
+                            .total_stake_amount
+                            .checked_sub(amount)
+                            .ok_or(ArithmeticError::Overflow)?;
+                        Ok(())
+                    })?;
+                    T::Assets::burn_from(Self::staking_currency()?, &Self::account_id(), amount)?;
+                }
+                XcmRequest::BondExtra { amount } if executed => {
+                    MatchingPool::<T>::try_mutate(|p| -> DispatchResult {
+                        p.total_stake_amount = p
+                            .total_stake_amount
+                            .checked_sub(amount)
+                            .ok_or(ArithmeticError::Overflow)?;
+                        Ok(())
+                    })?;
+                    T::Assets::burn_from(Self::staking_currency()?, &Self::account_id(), amount)?;
+                }
+                XcmRequest::Unbond { amount } if executed => {
+                    let liquid_amount = Self::exchange_rate()
```

Comment:
exchange_rate may change during this period

Reply:
yes but sometimes when there are many parachains, the xcm callback can be executed after many blocks I think

Reply:
like if polkadot is during auctions, crowdloans pallet has many XCM messages arriving in parallel chain's DMP queue. In this case the xcm callback may be executed much later

Reply:
emm... it seems like if we call settlement twice before xcm callback is executed, the exchange rate may be different. but we suppose to invoke `settlement` per era, which means 6 hours on Kusama and 1 day on Polkadot.

What do you think of this?

Reply:
do you mean it's impossible to not execute the callback after 6 hours?

Reply:
yes, I think so

Reply:
although yes it can be hard to happen but I think it's still preferred to calculate everything before. and store the result into XcmRequests, so that when we execute the callback we are sure nothing has changed it's just like an atomic operation

Reply:
yes exactly